ground/gcs/src/plugins/config/configautotunewidget.cpp

```diff
@@ -407,23 +395,309 @@ void ConfigAutotuneWidget::stuffShareForm(AutotuneFinalPage *autotuneShareForm)
+    for (int i = 0; i < fpts; i++) {
+        float x = delayed_fft[i];
+        float y = delayed_fft[i + fpts];
+        float u = orig_fft[i];
+        float v = orig_fft[i + fpts];
+
+        product[i] = -(u * x) - (v * y);
+        product[i + fpts] = (v * x) - (u * y);
+    }
+
+    /* Inverse FFT converts this to the time domain */
+    QVector<float> prod_time(pts);
+
+    fft.do_ifft(product.data(), prod_time.data());
+
+    /* And we take magnitudes to find tau. */
+    QVector<float> mags(fpts);
+
+    int max_idx = 0;
+    float max_val = 0;
+
+    for (int i = 0; i < fpts; i++) {
+        float real = prod_time[i];
+        float imag = prod_time[i + fpts];
+        mags[i] = sqrt(real * real + imag * imag);
+
+        if (mags[i] > max_val) {
+            max_val = mags[i];
+            max_idx = i;
+        }
+
+        // cout << i << "\t" << mags[i] << "\n";
+    }
+
+    // cout << "Maximum mags[" << max_idx << "] = " << max_val << "\n";
+
+    // TODO / optional: interpolate/find a better peak around max_idx.
+
+    return max_idx;
+}
+
+AutotunedValues ConfigAutotuneWidget::processAutotuneData(QByteArray *loadedFile)
+{
+    const at_flash *flash_data = reinterpret_cast<const at_flash *>((const void *)*loadedFile);
+
+    unsigned int size = loadedFile->size();
+
+    AutotunedValues vals = {};
+
+    /* Determine whether we have a sane amount of data, etc. */
+    if ((size < sizeof(at_flash)) || (flash_data->hdr.magic != ATFLASH_MAGIC)) {
+        /* XXX error dialog */
+        return vals;
+    }
+
+    unsigned int size_expected = sizeof(at_flash)
+        + sizeof(at_measurement) * flash_data->hdr.wiggle_points + flash_data->hdr.aux_data_len;
+
+    if (size < size_expected) {
+        /* XXX error dialog */
+        return vals;
+    }
+
+    float duration = (float)flash_data->hdr.wiggle_points / flash_data->hdr.sample_rate;
+
+    if ((duration < 0.25f) || (duration > 5.0f)) {
+        /* XXX error dialog */
+        return vals;
+    }
+
+    int pts = flash_data->hdr.wiggle_points;
+
+    for (int axis = 0; axis < 3; axis++) {
+        QVector<float> gyro_deriv(pts);
+        QVector<float> actu_desired(pts);
+
+        for (int i = 0; i < pts; i++) {
+            actu_desired[i] = flash_data->data[i].u[axis];
+        }
+
+        // Differentiate the gyro data
+        for (int i = 1; i < pts; i++) {
+            gyro_deriv[i] = flash_data->data[i].y[axis] - flash_data->data[i - 1].y[axis];
+        }
+
+        gyro_deriv[0] = flash_data->data[0].y[axis] - flash_data->data[pts - 1].y[axis];
+
+        float sample_tau = get_sample_delay(pts, gyro_deriv, actu_desired);
+
+        float tau = sample_tau / flash_data->hdr.sample_rate;
+
+        biquad_filter_in_place(1 / (sample_tau * M_PI * 1.414), pts, actu_desired.data());
+
+        QVector<float> gyro_sorted = gyro_deriv;
+        QVector<float> actu_sorted = actu_desired;
+
+        qSort(gyro_sorted.begin(), gyro_sorted.end());
```

Comment:
Use std::sort. qSort has been deprecated for some time http://doc.qt.io/qt-5/qtalgorithms.html#porting-guidelines.